File: General Problems/C++/merge_sort.cpp
#include <iostream>
#include <string>

using namespace std;

int* merge_sort(int* elements, int number_of_elements, bool reversed=false);
void print_list(int* elements, int number_of_elements, string name);

int main()
{
    int list[21] = {2, 5, 3, 1, 7, 4, 9, 8, 6, 15, 13, 18, 12, 11, 14, 17, 19, 20, 10, 16};
    int listLength = 20;
    int* orderedList = merge_sort(list, listLength);
    int* reversedOrderedList = merge_sort(list, listLength, true);
    print_list(list, listLength, "List");
    print_list(orderedList, listLength, "Ordered List");
    print_list(reversedOrderedList, listLength, "Reversed Ordered List");

}

/**
 * Sorts an array of integers using the merge sort algorithm.
 *
 * @param elements The array of integers to be sorted.
 * @param number_of_elements The number of elements in the array.
 * @param reversed A boolean indicating whether to sort in ascending (false) or descending (true) order.
 * @return A pointer to the sorted array.
 */
int* merge_sort(int* elements, int number_of_elements, bool reversed){
  // Base case: If there is only one element, return it as it is already sorted.
  if (number_of_elements == 1) {
    int* element = (int*) malloc(sizeof(int));
    *element = *elements;
    return element;
  } else {
    // Allocate memory for the sorted array.
    int* sorted_elements = (int*) malloc(sizeof(int) * number_of_elements);
    // Split the array into two halves.
    int* left_elements = (int*) malloc(sizeof(int) * (number_of_elements / 2));
    int* right_elements = (int*) malloc(sizeof(int) * (number_of_elements - number_of_elements / 2));

    // Recursively sort the left and right halves.
    left_elements = merge_sort(elements, number_of_elements / 2, reversed);
    right_elements = merge_sort(elements + number_of_elements / 2, number_of_elements - number_of_elements / 2, reversed);

    // Merge the sorted left and right halves.
    int left_index = 0;
    int right_index = 0;
    while (left_index + right_index < number_of_elements) {
      if (reversed) {
        // If sorting in descending order, compare elements in reverse.
        if (left_index < number_of_elements / 2 && right_index < number_of_elements - number_of_elements / 2) {
          if (*(left_elements + left_index) >= *(right_elements + right_index)) {
            *(sorted_elements + left_index + right_index) = *(left_elements + left_index);
            left_index++;
          } else {
            *(sorted_elements + left_index + right_index) = *(right_elements + right_index);
            right_index++;
          }
        } else {
          // If one half is exhausted, append the remaining elements from the other half.
          if (left_index == number_of_elements / 2) {
            *(sorted_elements + left_index + right_index) = *(right_elements + right_index);
            right_index++;
          } else {
            *(sorted_elements + left_index + right_index) = *(left_elements + left_index);
            left_index++;
          }
        }
      } else {
        // If sorting in ascending order, compare elements normally.
        if (left_index < number_of_elements / 2 && right_index < number_of_elements - number_of_elements / 2) {
          if (*(left_elements + left_index) <= *(right_elements + right_index)) {
            *(sorted_elements + left_index + right_index) = *(left_elements + left_index);
            left_index++;
          } else {
            *(sorted_elements + left_index + right_index) = *(right_elements + right_index);
            right_index++;
          }
        } else {
          // If one half is exhausted, append the remaining elements from the other half.
          if (left_index == number_of_elements / 2) {
            *(sorted_elements + left_index + right_index) = *(right_elements + right_index);
            right_index++;
          } else {
            *(sorted_elements + left_index + right_index) = *(left_elements + left_index);
            left_index++;
          }
        }
      }
    }
    return sorted_elements;
  }
}

/**
 * Prints an array of integers.
 *
 * @param elements The array of integers to be printed.
 * @param number_of_elements The number of elements in the array.
 * @param name The name to be printed before the array.
 */
void print_list(int* elements, int number_of_elements, string name) {
  cout << name << ": ";
  for (int index = 0; index < number_of_elements; index++)
    cout << elements[index] << " ";
  cout << endl;
}

File: General Problems/Python/merge_sort.py
def merge_sort(elements, reversed=False):
    """
    Sorts a list of integers using the merge sort algorithm.

    Args:
        elements (list): The list of integers to be sorted.
        reversed (bool, optional): A boolean indicating whether to sort in ascending (False) or descending (True) order. Defaults to False.

    Returns:
        list: The sorted list.
    """
    # Base case: If there is only one element, return it as it is already sorted.
    if len(elements) <= 1:
        return elements
    mid = len(elements) // 2
    # Recursively sort the left and right halves.
    left_half = merge_sort(elements[:mid], reversed)
    right_half = merge_sort(elements[mid:], reversed)
    # Merge the sorted left and right halves.
    return merge(left_half, right_half, reversed)

def merge(left, right, reversed):
    """
    Merges two sorted lists into one sorted list.

    Args:
        left (list): The first sorted list.
        right (list): The second sorted list.
        reversed (bool): A boolean indicating whether to sort in ascending (False) or descending (True) order.

    Returns:
        list: The merged sorted list.
    """
    merged = []
    left_index = 0
    right_index = 0
    while left_index < len(left) and right_index < len(right):
        if reversed:
            # If sorting in descending order, compare elements in reverse.
            if left[left_index] >= right[right_index]:
                merged.append(left[left_index])
                left_index += 1
            else:
                merged.append(right[right_index])
                right_index += 1
        else:
            # If sorting in ascending order, compare elements normally.
            if left[left_index] <= right[right_index]:
                merged.append(left[left_index])
                left_index += 1
            else:
                merged.append(right[right_index])
                right_index += 1
    # Append the remaining elements from both lists.
    merged += left[left_index:]
    merged += right[right_index:]
    return merged

def print_list(elements, name):
    """
    Prints a list of integers.

    Args:
        elements (list): The list of integers to be printed.
        name (str): The name to be printed before the list.
    """
    print(name + ": " + ' '.join(str(element) for element in elements))

list = [2, 5, 3, 1, 7, 4, 9, 8, 6, 15, 13, 18, 12, 11, 14, 17, 19, 20, 10, 16]
listLength = 20
orderedList = merge_sort(list)
reversedOrderedList = merge_sort(list, True)
print_list(list, "List")
print_list(orderedList, "Ordered List")
print_list(reversedOrderedList, "Reversed Ordered List")
